### ISSUE TITLE
prism and other features are not loaded/working inside a yarn workspace

### ISSUE THREAD

**Issue by dackmin** [facebook/docusaurus#5298]
### Prerequisites

- [x] I'm using the latest version of Docusaurus.
- [x] I have tried the `npm run clear` or `yarn clear` command.
- [x] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [x] I have tried creating a repro with https://new.docusaurus.io
- [x] I have read the console error message carefully (if applicable)

### Description

When using yarn workspaces, prism & probably some other markdown features (looking at you code block title) don't work as intended.

Normal project:
![Capture d’écran 2021-08-05 à 15 20 04](https://user-images.githubusercontent.com/1873323/128356851-5c85d04d-10e9-45aa-a1cd-170a73f149b0.png)

Same project inside a workspace:
![Capture d’écran 2021-08-05 à 15 20 18](https://user-images.githubusercontent.com/1873323/128356899-cf0d9545-ef2d-496a-a2f3-0992d38e692d.png)

### Have you read the [Contributing Guidelines on issues]
Yep.

## Steps to reproduce

- Fork the example provided by new.docusaurus.io
- Move everything to a subfolder, "custom" for example
- Create a "custom" yarn workspace:
```json
{
  "workspaces": ["custom"]
}
```
- and re-install/re-run the project

Repro can be found here: https://codesandbox.io/s/docusaurus-inside-workspace-9tn6n

## Expected behavior

Having wonderful code coloration ✨ 

## Actual behavior

Everything is dark and sad. And there's not even a single console line.

## Your environment

- Public site URL: https://codesandbox.io/s/docusaurus-inside-workspace-9tn6n
- Docusaurus version used: 2.0.0-beta.4 (but the problem was already here on 2.0.0-beta.0)
- Environment name and version: Chrome 92.0.4515.130, Node v14.16.1 & Yarn 1.22.10 on Code Sandbox, Node v16.6.0 & Yarn 1.22.5 on my own env
- Operating system and version: Debian 10 on Code Sandbox, macOS 11.5 on my own env

## Reproducible demo

https://codesandbox.io/s/docusaurus-inside-workspace-9tn6n

**Comment by slorber** [comment]
Hi,

I'm not sure using Codesandbox is a good fit to test yarn workspaces.

I created a local workspace with a brand new site and it's working for me, here's a zipfile that you can try running:

[monorepo-test.zip](https://github.com/facebook/docusaurus/files/6939787/monorepo-test.zip)

`yarn install && yarn dev`

![image](https://user-images.githubusercontent.com/749374/128377226-363cda98-e836-4bf8-b6ff-5cc5bf51d1a2.png)

--- 

It's hard to find out how you modified the sandbox exactly to see what could be wrong. If you still think there's a bug please provide a github repro where the first commit is the init template unmodified and you break it by transforming into a monorepo.

**Comment by dackmin** [comment]
The problem seems to occur when you lock the dependencies to the versions given by `npx @docusaurus/init@latest init`

```diff
diff --git a/my-website/package.json b/my-website/package.json
index 603681e..60b64ca 100644
--- a/my-website/package.json
+++ b/my-website/package.json
@@ -16,14 +16,14 @@
   "dependencies": {
     "@docusaurus/core": "2.0.0-beta.4",
     "@docusaurus/preset-classic": "2.0.0-beta.4",
-    "@mdx-js/react": "^1.6.21",
-    "@svgr/webpack": "^5.5.0",
-    "clsx": "^1.1.1",
-    "file-loader": "^6.2.0",
-    "prism-react-renderer": "^1.2.1",
-    "react": "^17.0.1",
-    "react-dom": "^17.0.1",
-    "url-loader": "^4.1.1"
+    "@mdx-js/react": "1.6.21",
+    "@svgr/webpack": "5.5.0",
+    "clsx": "1.1.1",
+    "file-loader": "6.2.0",
+    "prism-react-renderer": "1.2.1",
+    "react": "17.0.1",
+    "react-dom": "17.0.1",
+    "url-loader": "4.1.1"
   },
   "browserslist": {
     "production": [
@@ -37,4 +37,4 @@
       "last 1 safari version"
     ]
   }
-}
\ No newline at end of file
+}
diff --git a/yarn.lock b/yarn.lock
index dd16ad1..013b7be 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -1547,6 +1547,11 @@
     unist-builder "2.0.3"
     unist-util-visit "2.0.3"
 
+"@mdx-js/react@1.6.21":
+  version "1.6.21"
+  resolved "https://registry.yarnpkg.com/@mdx-js/react/-/react-1.6.21.tgz#86d962471a5e160c59a6b32054aa55c0c7ca404e"
+  integrity sha512-CgSNT9sq2LAlhEbVlPg7DwUQkypz+CWaWGcJbkgmp9WCAy6vW33CQ44UbKPiH3wet9o+UbXeQOqzZd041va83g==
+
 "@mdx-js/react@^1.6.21":
   version "1.6.22"
   resolved "https://registry.yarnpkg.com/@mdx-js/react/-/react-1.6.22.tgz#ae09b4744fddc74714ee9f9d6f17a66e77c43573"
@@ -1705,7 +1710,7 @@
     deepmerge "^4.2.2"
     svgo "^1.2.2"
 
-"@svgr/webpack@^5.5.0":
+"@svgr/webpack@5.5.0", "@svgr/webpack@^5.5.0":
   version "5.5.0"
   resolved "https://registry.yarnpkg.com/@svgr/webpack/-/webpack-5.5.0.tgz#aae858ee579f5fa8ce6c3166ef56c6a1b381b640"
   integrity sha512-DOBOK255wfQxguUta2INKkzPj6AIS6iafZYiYmHn6W3pHlycSRRlvWKCfLDG10fXfLWqE3DJHgRUOyJYmARa7g==
@@ -2722,7 +2727,7 @@ clone-response@^1.0.2:
   dependencies:
     mimic-response "^1.0.0"
 
-clsx@^1.1.1:
+clsx@1.1.1, clsx@^1.1.1:
   version "1.1.1"
   resolved "https://registry.yarnpkg.com/clsx/-/clsx-1.1.1.tgz#98b3134f9abbdf23b2663491ace13c5c03a73188"
   integrity sha512-6/bPho624p3S2pMyvP5kKBPXnI3ufHLObBFCfgx+LkeR5lg2XYy2hqZqUf45ypD8COn2bhgGJSUE+l5dhNBieA==
@@ -3897,7 +3902,7 @@ figures@^3.2.0:
   dependencies:
     escape-string-regexp "^1.0.5"
 
-file-loader@^6.2.0:
+file-loader@6.2.0, file-loader@^6.2.0:
   version "6.2.0"
   resolved "https://registry.yarnpkg.com/file-loader/-/file-loader-6.2.0.tgz#baef7cf8e1840df325e4390b4484879480eebe4d"
   integrity sha512-qo3glqyTa61Ytg4u73GultjHGjdRyig3tG6lPtyX/jOEJvHif9uB0/OCI2Kif6ctF3caQTW2G5gym21oAsI4pw==
@@ -6597,7 +6602,7 @@ pretty-time@^1.1.0:
   resolved "https://registry.yarnpkg.com/pretty-time/-/pretty-time-1.1.0.tgz#ffb7429afabb8535c346a34e41873adf3d74dd0e"
   integrity sha512-28iF6xPQrP8Oa6uxE6a1biz+lWeTOAPKggvjB8HAs6nVMKZwf5bG++632Dx614hIWgUPkgivRfG+a8uAXGTIbA==
 
-prism-react-renderer@^1.2.1:
+prism-react-renderer@1.2.1, prism-react-renderer@^1.2.1:
   version "1.2.1"
   resolved "https://registry.yarnpkg.com/prism-react-renderer/-/prism-react-renderer-1.2.1.tgz#392460acf63540960e5e3caa699d851264e99b89"
   integrity sha512-w23ch4f75V1Tnz8DajsYKvY5lF7H1+WvzvLUcF0paFxkTHSp42RS0H5CttdN2Q8RR3DRGZ9v5xD/h3n8C8kGmg==
@@ -6801,14 +6806,14 @@ react-dev-utils@^11.0.1:
     strip-ansi "6.0.0"
     text-table "0.2.0"
 
-react-dom@^17.0.1:
-  version "17.0.2"
-  resolved "https://registry.yarnpkg.com/react-dom/-/react-dom-17.0.2.tgz#ecffb6845e3ad8dbfcdc498f0d0a939736502c23"
-  integrity sha512-s4h96KtLDUQlsENhMn1ar8t2bEa+q/YAtj8pPPdIjPDGBDIVNsrD9aXNWqspUe6AzKCIG0C1HZZLqLV7qpOBGA==
+react-dom@17.0.1:
+  version "17.0.1"
+  resolved "https://registry.yarnpkg.com/react-dom/-/react-dom-17.0.1.tgz#1de2560474ec9f0e334285662ede52dbc5426fc6"
+  integrity sha512-6eV150oJZ9U2t9svnsspTMrWNyHc6chX0KzDeAOXftRa8bNeOKTTfCJ7KorIwenkHd2xqVTBTCZd79yk/lx/Ug==
   dependencies:
     loose-envify "^1.1.0"
     object-assign "^4.1.1"
-    scheduler "^0.20.2"
+    scheduler "^0.20.1"
 
 react-error-overlay@^6.0.9:
   version "6.0.9"
@@ -6914,10 +6919,10 @@ react-textarea-autosize@^8.3.2:
     use-composed-ref "^1.0.0"
     use-latest "^1.0.0"
 
-react@^17.0.1:
-  version "17.0.2"
-  resolved "https://registry.yarnpkg.com/react/-/react-17.0.2.tgz#d0b5cc516d29eb3eee383f75b62864cfb6800037"
-  integrity sha512-gnhPt75i/dq/z3/6q/0asP78D0u592D5L1pd7M8P+dck6Fu/jJeL6iVVK23fptSUZj8Vjf++7wXA8UNclGQcbA==
+react@17.0.1:
+  version "17.0.1"
+  resolved "https://registry.yarnpkg.com/react/-/react-17.0.1.tgz#6e0600416bd57574e3f86d92edba3d9008726127"
+  integrity sha512-lG9c9UuMHdcAexXtigOZLX8exLWkW0Ku29qPRU8uhF2R9BN96dLCt0psvzPLlHc5OWkgymP3qwTRgbnw5BKx3w==
   dependencies:
     loose-envify "^1.1.0"
     object-assign "^4.1.1"
@@ -7321,7 +7326,7 @@ sax@^1.2.4, sax@~1.2.4:
   resolved "https://registry.yarnpkg.com/sax/-/sax-1.2.4.tgz#2816234e2378bddc4e5354fab5caa895df7100d9"
   integrity sha512-NqVDv9TpANUjFm0N8uM5GxL36UgKi9/atZw+x7YFnQ8ckwFGKrl4xX4yWtrey3UJm5nP1kUbnYgLopqWNSRhWw==
 
-scheduler@^0.20.2:
+scheduler@^0.20.1:
   version "0.20.2"
   resolved "https://registry.yarnpkg.com/scheduler/-/scheduler-0.20.2.tgz#4baee39436e34aa93b4874bddcbf0fe8b8b50e91"
   integrity sha512-2eWfGgAqqWFGqtdMmcL5zCMK1U8KlXv8SQFGglL3CEtd0aDVDWgeF/YoCmvln55m5zSk3J/20hTaSBeSObsQDQ==
@@ -8325,7 +8330,7 @@ urix@^0.1.0:
   resolved "https://registry.yarnpkg.com/urix/-/urix-0.1.0.tgz#da937f7a62e21fec1fd18d49b35c2935067a6c72"
   integrity sha1-2pN/emLiH+wf0Y1Js1wpNQZ6bHI=
 
-url-loader@^4.1.1:
+url-loader@4.1.1, url-loader@^4.1.1:
   version "4.1.1"
   resolved "https://registry.yarnpkg.com/url-loader/-/url-loader-4.1.1.tgz#28505e905cae158cf07c92ca622d7f237e70a4e2"
   integrity sha512-3BTV812+AVHHOJQO8O5MkWgZ5aosP7GnROJwvzLS9hWDj00lZ6Z0wNak423Lp9PBZN05N+Jk/N5Si8jRAlGyWA==
```

and is fixed by updating `@mdx-js/react` to 1.16.22.

Here's the repo with the commit history: https://github.com/dackmin/docusaurus-issue.
Btw, the repro I gave on CodeSandbox is working fine with 1.16.22 too, so I don't see the problem with a CodeSandbox repro even for yarn workspaces.

------

On a side note, closing an issue before even remotely trying to understand what's wrong does not benefit anyone.

**Comment by slorber** [comment]
> Btw, the repro I gave on CodeSandbox is working fine with 1.16.22 too, so I don't see the problem with a CodeSandbox repro even for yarn workspaces.

CodeSandbox won't tell me easily what you modified exactly, that's why I think in this case it's not a good fit, because I'm not willing to invest a lot of time to troubleshoot a problem that may be caused by a typo somewhere. 

It would have been time-consuming for me to figure out that you decided to lock those versions instead of using the ^ semver ranges, as it's not mentioned in the repro steps.

> On a side note, closing an issue before even remotely trying to understand what's wrong does not benefit anyone.

Isn't investing my time to init a new site and run your wrong repro steps locally, figuring out that I can't reproduce with those steps enough? 

I'm always open to further discussions and happy to re-open an issue if I'm proven wrong, but also try to proactively close issues that I think are not bugs in Docusaurus because reporters tend to move away and issues stay open forever.

Bugs in transitive dependencies happen all the time and not locking to a specific version might help pick a fix. Just today a little bug in Babel lead to many issues in Docusaurus, React-Native, Expo, Gatsby... All fixed by just re-generating the lockfile a few hours later.

**Comment by dackmin** [comment]
Alright thanks!

Do you think it might be good to update `@mdx-js/react` to 1.16.22 inside the template generated by `init` or it is probably better to assume it won't happen again due to generally non-locked dependencies or managers like renovate doing their job?

**Comment by slorber** [comment]
Updating the min version make sense so that it doesn't happen again more reliably, but are we 100% sure it's a mdxjs version issue? They are other changes in this lock file. 

My instinct is that the problem might also be related to using 2 different versions at the same time. 

To be sure I'd suggest first reducing that diff to a single version change, using yarn resolution if needed to avoid duplicates.

**Comment by dackmin** [comment]
Yep sorry, the previous diff is the broken one. This one is the one fixing everything:

```diff
diff --git a/my-website/package.json b/my-website/package.json
index 60b64ca..8f7518b 100644
--- a/my-website/package.json
+++ b/my-website/package.json
@@ -16,7 +16,7 @@
   "dependencies": {
     "@docusaurus/core": "2.0.0-beta.4",
     "@docusaurus/preset-classic": "2.0.0-beta.4",
-    "@mdx-js/react": "1.6.21",
+    "@mdx-js/react": "1.6.22",
     "@svgr/webpack": "5.5.0",
     "clsx": "1.1.1",
     "file-loader": "6.2.0",
diff --git a/yarn.lock b/yarn.lock
index 013b7be..29a97b8 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -1547,12 +1547,7 @@
     unist-builder "2.0.3"
     unist-util-visit "2.0.3"
 
-"@mdx-js/react@1.6.21":
-  version "1.6.21"
-  resolved "https://registry.yarnpkg.com/@mdx-js/react/-/react-1.6.21.tgz#86d962471a5e160c59a6b32054aa55c0c7ca404e"
-  integrity sha512-CgSNT9sq2LAlhEbVlPg7DwUQkypz+CWaWGcJbkgmp9WCAy6vW33CQ44UbKPiH3wet9o+UbXeQOqzZd041va83g==
-
-"@mdx-js/react@^1.6.21":
+"@mdx-js/react@1.6.22", "@mdx-js/react@^1.6.21":
   version "1.6.22"
   resolved "https://registry.yarnpkg.com/@mdx-js/react/-/react-1.6.22.tgz#ae09b4744fddc74714ee9f9d6f17a66e77c43573"
   integrity sha512-TDoPum4SHdfPiGSAaRBw7ECyI8VaHpK8GJugbJIJuqyh6kzw9ZLJZW3HGL3NNrJGxcAixUvqROm+YuQOo5eXtg==
```

Corresponding commit on the repro repo: https://github.com/dackmin/docusaurus-issue/commit/f28688975bfe2b2ca774fe308e40d67bf95d308b

v1.16.21 is also used here:
https://github.com/facebook/docusaurus/tree/master/packages/docusaurus-mdx-loader & https://github.com/facebook/docusaurus/tree/master/packages/docusaurus-theme-classic

**Comment by slorber** [comment]
hey

to me, the problem is not about 1.6.21, but it's about having 2 versions of this package in your repo, one hoisted in `node_modules` and one in `my-website/node_modules`.

Using `1.6.21` is fine if you make sure that your monorepo only install this package once.

You can test this easily by adding this at the top of your monorepo and see that it still works despite not using the latest version:

```js
  "resolutions": {
    "@mdx-js/react": "1.6.21"
  }
```

So basically, updating the version does not solve the problem, it's making sure you don't install twice the same package because different parts of your Docusaurus project (your code vs internal code) will use different mdx package versions.


I'm surprised that Yarn isn't able to deduplicate this better though, it could figure out a way to hoist only one package version, as all deps ranges are compatible
